### PR TITLE
Additional functionality for tests

### PR DESCRIPTION
**Pull request by atsikham**
Added additional things such as:
* ENIs detach/remove for safe subnets deletion
* EKS resources deletion (eks, node group, iam, cloudwatch log group)

I wasn't able to fix an issue with ENIs, but found such issue that can be related - https://github.com/aws/amazon-vpc-cni-k8s/issues/1223, probably we need a workaround to handle ENIs related to k8s nodes.

### CODE REVIEW

File: integration_test.go [hunk 1]
```diff
@@ -953,3 +1028,239 @@ func removeResourceGroup(t *testing.T, session *session.Session, rgToRemoveName
...
+            case eks.ErrCodeClientException:
+                t.Fatal(eks.ErrCodeResourceInUseException, aerr.Error())
+            case eks.ErrCodeServerException:
+                t.Fatal(eks.ErrCodeResourceInUseException, aerr.Error())
+            case eks.ErrCodeServiceUnavailableException:
+                t.Fatal(eks.ErrCodeResourceInUseException, aerr.Error())
+            case eks.ErrCodeInvalidParameterException:
+                t.Fatal(eks.ErrCodeInvalidParameterException, aerr.Error())
+            default:
+                fmt.Println(aerr.Error())
```
Comment (seriva) [inline integration_test.go:1258]: No t.Fatal here?

File: integration_test.go [hunk 2]
```diff
@@ -953,3 +1028,239 @@ func removeResourceGroup(t *testing.T, session *session.Session, rgToRemoveName
...
+
+func removeCluster(t *testing.T, session *session.Session, clusterName string) {
+    eksClient := eks.New(session)
+    eksDelIn := &eks.DeleteClusterInput{
+        Name: aws.String(clusterName),
+    }
+    _, err := eksClient.DeleteCluster(eksDelIn)
+    if err != nil {
+        if aerr, ok := err.(awserr.Error); ok {
+            switch aerr.Code() {
```
Comment (seriva) [inline integration_test.go:1213]: This is a general comment about all these error handling switch blockes inside this review. Just wondering if the level of handling like thatis needed? Maybe just something like:

```Go
 if aerr, ok := err.(awserr.Error); ok {
    t.Fatal(aerr.Code(), aerr.Error())
   ....
```

File: resources/terraform/infra.tf [hunk 3]
```diff
@@ -18,7 +18,7 @@ resource "aws_subnet" "eks_subnet" {
   cidr_block        = cidrsubnet(data.aws_vpc.vpc.cidr_block, 4, 15-count.index)
   vpc_id            = data.aws_vpc.vpc.id
   tags              = {
-    name                                = "${var.name}-eks-subnet${count.index}"
+    Name                                = "${var.name}-eks-subnet${count.index}"
```
Comment (seriva) [inline resources/terraform/infra.tf:21]: Any issues with this when its lowercase?

File: integration_test.go [hunk 4]
```diff
@@ -953,3 +1028,239 @@ func removeResourceGroup(t *testing.T, session *session.Session, rgToRemoveName
...
+            case eks.ErrCodeResourceInUseException:
+                t.Fatal(eks.ErrCodeResourceInUseException, aerr.Error())
+            case eks.ErrCodeClientException:
+                t.Fatal(eks.ErrCodeResourceInUseException, aerr.Error())
+            case eks.ErrCodeServerException:
+                t.Fatal(eks.ErrCodeResourceInUseException, aerr.Error())
+            case eks.ErrCodeServiceUnavailableException:
+                t.Fatal(eks.ErrCodeResourceInUseException, aerr.Error())
+            default:
+                fmt.Println(aerr.Error())
```
Comment (seriva) [inline integration_test.go:1225]: No t.Fatal here?

File: resources/terraform/infra.tf (hunk 3 above)
Comment (atsikham) [reply resources/terraform/infra.tf:21]: Yes, resource name depends on this tag. When lowercase is used, there is no resource name displayed.

File: integration_test.go (hunk 2 above)
Comment (atsikham) [reply integration_test.go:1213]: This is how it's done in API usage [examples](https://docs.aws.amazon.com/sdk-for-go/api/service/eks/#EKS.DeleteCluster), not for all API there is an example, but in each existing it's so. So final decision is to change that?

File: integration_test.go (hunk 1 above)
Comment (atsikham) [reply integration_test.go]: Fixed.

File: integration_test.go (hunk 4 above)
Comment (atsikham) [reply integration_test.go]: Fixed.

File: integration_test.go (hunk 2 above)
Comment (seriva) [reply integration_test.go:1213]: For me it doesnt something like this just doesnt make much sence:

```Go
	    case eks.ErrCodeResourceInUseException:
                t.Fatal(eks.ErrCodeResourceInUseException, aerr.Error())
            case eks.ErrCodeClientException:
                t.Fatal(eks.ErrCodeResourceInUseException, aerr.Error())
            case eks.ErrCodeServerException:
                t.Fatal(eks.ErrCodeResourceInUseException, aerr.Error())
            case eks.ErrCodeServiceUnavailableException:
                t.Fatal(eks.ErrCodeResourceInUseException, aerr.Error())
```
I would at least just put this under default with:

```Go
t.Fatal(aerr.Code(), aerr.Error())
```

File: integration_test.go [hunk 5]
```diff
@@ -953,3 +1031,177 @@ func removeResourceGroup(t *testing.T, session *session.Session, rgToRemoveName
...
+
+			policyDetachIn := &iam.DetachRolePolicyInput{
+				PolicyArn: aws.String(*policy.PolicyArn),
+				RoleName:  aws.String(roleName),
+			}
+
+			_, errDetach := iamClient.DetachRolePolicy(policyDetachIn)
+			if errDetach != nil {
+				if aerr, ok := errDetach.(awserr.Error); ok {
+					switch aerr.Code() {
```
Comment (erzetpe) [inline integration_test.go:1073]: This looks weird. why like that?

File: integration_test.go [hunk 6]
```diff
@@ -953,3 +1031,177 @@ func removeResourceGroup(t *testing.T, session *session.Session, rgToRemoveName
...
+
+		roleInlineListIn := &iam.ListRolePoliciesInput{
+			RoleName: aws.String(roleName),
+		}
+
+		// List inline policies for role
+		inlinePolicies, errPolicyInlineList := iamClient.ListRolePolicies(roleInlineListIn)
+		if errPolicyInlineList != nil {
+			if aerr, ok := errPolicyInlineList.(awserr.Error); ok {
+				switch aerr.Code() {
```
Comment (erzetpe) [inline integration_test.go:1091]: Same here.

File: integration_test.go [hunk 7]
```diff
@@ -953,3 +1031,177 @@ func removeResourceGroup(t *testing.T, session *session.Session, rgToRemoveName
...
+
+		roleListIn := &iam.ListAttachedRolePoliciesInput{
+			RoleName: aws.String(roleName),
+		}
+
+		// List managed policies for role
+		policies, errPolicyList := iamClient.ListAttachedRolePolicies(roleListIn)
+		if errPolicyList != nil {
+			if aerr, ok := errPolicyList.(awserr.Error); ok {
+				switch aerr.Code() {
```
Comment (erzetpe) [inline integration_test.go:1050]: Maybe easier here will be to use if and switch is an overkill?

File: integration_test.go [hunk 8]
```diff
@@ -953,3 +1031,177 @@ func removeResourceGroup(t *testing.T, session *session.Session, rgToRemoveName
...
+
+			policyDeleteIn := &iam.DeleteRolePolicyInput{
+				PolicyName: aws.String(*inlinePolicy),
+				RoleName:   aws.String(roleName),
+			}
+
+			_, errDelete := iamClient.DeleteRolePolicy(policyDeleteIn)
+			if errDelete != nil {
+				if aerr, ok := errDelete.(awserr.Error); ok {
+					switch aerr.Code() {
```
Comment (erzetpe) [inline integration_test.go:1111]: Same here.

File: integration_test.go [hunk 9]
```diff
@@ -953,3 +1031,177 @@ func removeResourceGroup(t *testing.T, session *session.Session, rgToRemoveName
...
+
+		roleDelIn := &iam.DeleteRoleInput{
+			RoleName: aws.String(roleName),
+		}
+
+		// Delete role
+		_, errDeleteRole := iamClient.DeleteRole(roleDelIn)
+		if errDeleteRole != nil {
+			if aerr, ok := errDeleteRole.(awserr.Error); ok {
+				switch aerr.Code() {
```
Comment (erzetpe) [inline integration_test.go:1129]: Same here.

File: integration_test.go [hunk 10]
```diff
@@ -953,3 +1031,177 @@ func removeResourceGroup(t *testing.T, session *session.Session, rgToRemoveName
...
+func removeLogGroup(t *testing.T, session *session.Session, groupName string) {
+	logsClient := cloudwatchlogs.New(session)
+	logGroupDelIn := &cloudwatchlogs.DeleteLogGroupInput{
+		LogGroupName: aws.String(groupName),
+	}
+	_, errLogGroupDel := logsClient.DeleteLogGroup(logGroupDelIn)
+	if errLogGroupDel != nil {
+		if aerr, ok := errLogGroupDel.(awserr.Error); ok {
+			switch aerr.Code() {
+			case cloudwatchlogs.ErrCodeResourceNotFoundException:
```
Comment (erzetpe) [inline integration_test.go:1151]: Maybe here will be better to use if?

File: integration_test.go [hunk 11]
```diff
@@ -953,3 +1031,177 @@ func removeResourceGroup(t *testing.T, session *session.Session, rgToRemoveName
...
+
+func removeCluster(t *testing.T, session *session.Session, clusterName string) {
+	eksClient := eks.New(session)
+	eksDelIn := &eks.DeleteClusterInput{
+		Name: aws.String(clusterName),
+	}
+	_, err := eksClient.DeleteCluster(eksDelIn)
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			switch aerr.Code() {
```
Comment (erzetpe) [inline integration_test.go:1172]: Same here.

File: integration_test.go [hunk 12]
```diff
@@ -953,3 +1031,177 @@ func removeResourceGroup(t *testing.T, session *session.Session, rgToRemoveName
...
+func removeNodeGroup(t *testing.T, session *session.Session, clusterName string, nodeGroupName string) {
+	eksClient := eks.New(session)
+	ngDelIn := &eks.DeleteNodegroupInput{
+		ClusterName:   aws.String(clusterName),
+		NodegroupName: aws.String(nodeGroupName),
+	}
+	_, err := eksClient.DeleteNodegroup(ngDelIn)
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			switch aerr.Code() {
```
Comment (erzetpe) [inline integration_test.go:1195]: Same here.

**Review by erzetpe — CHANGES_REQUESTED** [review]

**Review by erzetpe — APPROVED** [review]

**Comment by mkyc** [comment]
@atsikham please connect this PR to issue in ZH.

**Comment by atsikham** [comment]
@mkyc done

**Review by seriva — APPROVED** [review]

**Pull request merged**